src/java/org/apache/cassandra/utils/concurrent/SkipListProgressWaitQueue.java

```diff
@@ -0,0 +1,183 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.utils.concurrent;
+
+import java.util.Iterator;
+import java.util.concurrent.ConcurrentNavigableMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.concurrent.locks.LockSupport;
+
+import com.google.common.collect.ImmutableMap;
+
+import com.codahale.metrics.Timer;
+import com.palantir.tracing.CloseableTracer;
+import com.palantir.tracing.DetachedSpan;
+import com.palantir.tracing.Tracer;
+
+public final class SkipListProgressWaitQueue implements ProgressWaitQueue
+{
+    private static final int CANCELLED = -1;
+    private static final int SIGNALLED = 1;
+    private static final int NOT_SET = 0;
+
+    private static final AtomicIntegerFieldUpdater signalledUpdater = AtomicIntegerFieldUpdater.newUpdater(SkipListProgressWaitQueue.RegisteredSignal.class, "state");
+
+    private final ConcurrentNavigableMap<Long, RegisteredSignal> waiters = new ConcurrentSkipListMap<>();
+
+    public WaitQueue.Signal register(long waitUntil)
+    {
+        RegisteredSignal signal = new RegisteredSignal();
+        waiters.put(waitUntil, signal);
+        return signal;
+    }
+
+    public WaitQueue.Signal register(long waitUntil, Timer.Context context)
+    {
+        assert context != null;
+        RegisteredSignal signal = new TimedSignal(context);
+        waiters.put(waitUntil, signal);
+        return signal;
+    }
+
+    public void signalUntil(long until) {
+        ConcurrentNavigableMap<Long, RegisteredSignal> dueWaiters = waiters.headMap(until, true);
+        try (CloseableTracer ignored = CloseableTracer.startSpan(
+            "SkipListProgressWaitQueue#signalAll",
+                ImmutableMap.of(
+                    "numWaiters", Integer.toString(waiters.size()),
+                    "numDueWaiters", Integer.toString(dueWaiters.size()))))
+        {
+            Iterator<RegisteredSignal> iter = dueWaiters.values().iterator();
+            while (iter.hasNext())
+            {
+                RegisteredSignal signal = iter.next();
+                signal.signal();
+                iter.remove();
+            }
+        }
+    }
+
+    private void cleanUpCancelled()
+    {
+        waiters.values().removeIf(RegisteredSignal::isCancelled);
+    }
+
+    /**
+     * A signal registered with this WaitQueue
+     */
+    private class RegisteredSignal extends WaitQueue.AbstractSignal
```

Comment:
A couple of thoughts/questions here:
- should we be closing these spans in the non-signaling case? and should we include that metadata in the span?
- it looks like the only differences between this and the `WaitQueue.RegisteredSignal` class is the tracing... should we just combine those? (looks like there may be a visibility issue with `signal()`, which could maybe be package private?)

Reply:
Handled the first point. 👍